### PR TITLE
specifies the encoding of path and query on get_uri.

### PR DESCRIPTION
**Pull request by jcchavezs**
Comes from https://github.com/mosn/mosn/pull/2168/files/c1e149b8720d28ebf4910be2f1ca03ca3e005093#r1015930151

### CODE REVIEW

File: content/http-handler-abi.md [hunk 1]
```diff
@@ -767,7 +779,8 @@ method to "POST".
 ;; get_uri writes the URI to memory if it isn't larger than `buf_limit`, e.g.
 ;; "/v1.0/hi?name=panda". The result is its length in bytes.  
 ;;
-;; Note: The URI may include query parameters.
+;; Note: The URI may include query parameters. Despite the path and query being
```
Comment (jcchavezs) [inline content/http-handler-abi.md:782]: This is the important change.

File: content/http-handler-abi.md [hunk 2]
```diff
@@ -156,10 +158,10 @@ This parameter supports the most common case of retrieving a header value by
 name. However, there are some subtle use cases possible, particularly helpful
 for WebAssembly performance:
 
-- re-using a buffer for reading header or path values (`buf`).
-- growing a buffer only when needed (retry with larger `buf_limit`).
-- avoiding copying invalidly large header values (`buf_limit`).
-- determining if a header exists without copying it (`buf_limit=0`).
+* re-using a buffer for reading header or path values (`buf`).
```
Comment (jcchavezs) [inline content/http-handler-abi.md:161]: This was autogenerated by IDE I guess to keep consistency with other lists (e.g. line 90). Can rollback if you prefer so @codefromthecrypt

**Review by codefromthecrypt — COMMENTED** [review]
I'm ok with this if we have a good reason for it, especially as it is a behavior change. Is this to help port other ABI to http-wasm? I know some don't actually support URI, but encode it in :host pseudo header, though not really aware the rules of it. Mainly, when changing we should add rationale for the decision and be aware of impact. Mind checking around?

cc
  @taoyuanyuan @antJack (mosn)
cc
  @ItalyPaleAle @daixiang0 (dapr)

To help share the load, I found some links of various ABI or context we could use for RATIONALE. Of course one way is to document this as encoded (change the impl to actually do that), and be prepared for a request for raw_uri later. Hope this helps and thanks for owning this change @jcchavezs! Regardless of choice we should document it, implement it, and be consistent.

https://github.com/fastly/Viceroy/blob/48ea9efc79ec8cd9b73a792559c561338c8c26ad/lib/src/wiggle_abi/req_impl.rs#L434-L461
https://github.com/deislabs/wagi/blob/c48bdd6a205920cf90878c558b60fc667289431a/src/http_util.rs#L117-L146
https://github.com/edgedb/edgedb-wasm/blob/74cd1196a551c666447cb45b9fec656176bb2cde/sdk/src/http_server.rs#L15
https://www.rfc-editor.org/rfc/rfc7540

File: content/http-handler-abi.md (hunk 1 above)
Comment (anuraaga) [reply content/http-handler-abi.md:782]: BTW I'd also lean towards non-encoded for URI here, it's generally what's used inside middleware like a routing one. Either way, guest SDKs could provide both but the ABI could focus on the commonly used case I guess.

Comment (codefromthecrypt) [reply content/http-handler-abi.md:782]: I would also like to clarify that even if I suggested we could offer two abi to get headers I really hope we don't because decisions like this could explode the ABI. In other words, a small ABI is a feature and there should be a very good reason for each new function to exist.

For starters, I'll make the impl consistent on raw and hopefully we can re-use this PR to document that and why. 

Here's a pitch on RATIONALE based on @anuraaga's note and some things I was thinking about similarly.

URI is returned raw because that's what's used in routing. An http-handler is an inbound processor, and HTTP properties such as HTTP path and query are often used for routing properties. If we returned the URI decoded, the guest would have higher overhead decoding these in order to route based on them.

Moreover, returning the URI raw from `get_uri` matches that we accept it raw in `set_uri`. Even if a host chooses to coerce an encoded input to `set_uri`, having the requirement of input matching output in the specified case is easier to reason with than otherwise.

Comment (anuraaga) [reply content/http-handler-abi.md:782]: Tried writing some code using getURI and realized this probably does need to be encoded, and maybe what JC had in mind. e.g., if getting `/disney?name=chip&dale` this would be two separate parameters when unencoded though most likely the encoded version was `/disney?name=chip%26dale`. Having raw values seems to only make sense when splitting the path and query but not when they're together because of the semantic meaning of `&`.

So I guess we can let `get_uri` and `set_uri` in the ABI deal with encoded, and let guest SDKs expose something easier, e.g. in tinygo SDK maybe we can replace `GetURI() string` with `GetURI() URL`. Haven't completely vetted this idea yet

Comment (codefromthecrypt) [reply content/http-handler-abi.md:782]: ok sounds like progress towards rationale.

I guess if we are going to encode the query, we'd do that with the query encoding which I always forget has slightly different rules than url. For years, I look at okhttp for some cheat sheet examples..

https://github.com/square/okhttp/blob/58ee1ce170c2aea75a113012956260873e808dad/okhttp/src/jvmMain/kotlin/okhttp3/HttpUrl.kt

**Comment by anuraaga** [comment]
I have run some tests with get-uri/set-uri with encoding against expressjs, mosn, nethttp (which I think implies Dapr which just reuses the nethttp middleware as is). It would be good to get a double check that these are the cases that we are concerned with

https://github.com/anuraaga/http-wasm-tck/blob/main/run.go#L82
https://github.com/anuraaga/http-wasm-tck/blob/main/guest/main.go#L59

The encoded URIs are passed between the ABI as is as expected, and reflects how the ABI is basically just a view on the raw HTTP data and nothing particularly semantic. 

We can add APIs in the guests that accept decoded uris to make easier to use, a start for tinygo is probably to have `setURI` accept `net.URL` intead of `string`. 

@jcchavezs would be good to doublecheck if this all checks out.

**Comment by codefromthecrypt** [comment]
sounds good folks. I think that a specification that tells you the intent (raw encoding) and suggestions on how to not goof that (using language SDK for URI) is a good thing. Particularly an SDK is a fantastic improvement of status quo! I'll leave it to you to merge on this approach

File: content/http-handler-abi.md (hunk 1 above)
Comment (codefromthecrypt) [reply content/http-handler-abi.md:782]: I think recommending types is not a good idea due to how compilers bloat things and also unnecessary overhead. 

Remember in many languages there are alternative HTTP libraries and they use things like buffers or strings to hydrate their URI types as well. Some may want to use a nonstd library and still have to suffer through the bloat using whatever the URI is (especially Java's is bad).

Comment (codefromthecrypt) [reply content/http-handler-abi.md:782]: case in point is the WIP tinygo thing bloated usage of get_uri by 89KB (1200%)

I think the thinking is better to be lean on types by default as it keeps nervous types like me from having to keep watching to make sure the thing doesn't go sideways.

File: content/rationale.md [hunk 3]
```diff
@@ -77,6 +77,15 @@ the module name and retains function names in the case format they are declared
 in. Parameters names are inconsistently mapped, preferring the case format of
 the language.
 
+## Why are the query parameters URI encoded?
```
Comment (jcchavezs) [inline content/rationale.md:80]: How about this @codefromthecrypt ?

Comment (codefromthecrypt) [reply content/rationale.md:80]: looks good. only thing left is clarifying why we do this host side.

Ex. it isn't really the user who is calling set_uri, it is the guest code. so, the guest SDKs are allowed to not encode the inputs and let the host do it instead. The host MUST ensure a raw query is encoded to prevent the issue you are mentioning. The reason why we allow the guest to opt-out is to keep their code size down. If we instead said the guest must encode, their wasm would bloat with logic to perform encoding, possibly even allocating, resulting in a more complex minimal guest.

TL;DR; this is right about why we need to encode, and we can also add that push this to the host to make guest SDKs easier and leaner. e.g. without the host doing this, it would be near impractical to implement anything in `%.wat`, and fuller featured languages like tinygo would both have more code and higher overhead to attempt to encode first.

Comment (anuraaga) [reply content/rationale.md:80]: AFAIK, both get_uri and set_uri have the same ambiguity issue, `set_uri("/disney?name=chip&dale")`, with the intention of sending two query parameters, `name` and `dale`, if the host encodes this it will become `set_uri("/disney?name=chip%26dale")` with just one parameter `name`.

I don't see a way around this without expecting the guest to pass an encoded URL. That doesn't mean the guest SDK has to encode, hence https://github.com/http-wasm/http-wasm-guest-tinygo/pull/22, and the user could use hard-coded encoded strings to avoid code or for use in `%.wat` if that's viable. Logic acting on non-hardcoded strings will likely need to bring in the code though.

On the bright side, none of our current hosts encode a passed in URI from what I can tell, for example we parse an encoded URI here

https://github.com/http-wasm/http-wasm-host-go/blob/main/handler/nethttp/host.go#L67

File: content/rationale.md [hunk 4]
```diff
@@ -77,6 +77,15 @@ the module name and retains function names in the case format they are declared
...
+
+Query parameters can contain characters that act as values or as a separator in the syntax. A query string like `?name=chip&dale` can represent two things depending if it was encoded or not:
+
+* encoded: `name` equals `chip` and `dale` is empty.
+* raw: `name` equals `chip&dale`
+
+To remove this ambiguity we will always return the URI encoded to reflect the query parameters intent. For example, if user wants to pass `chip&dale` as `name` the URI will be `/disney?name=chip%26dale` whereas if it is the case where `name` is `chip` and an additional query parameter called `dale` as empty the URI would be `/disney?name=chip&dale`.
```
Comment (anuraaga) [inline content/rationale.md:87]:
```suggestion
To remove this ambiguity we will always treat URIs returned from or passed to the ABI as encoded to reflect the query parameters intent. For example, if user wants to pass `chip&dale` as `name` the URI will be `/disney?name=chip%26dale` whereas if it is the case where `name` is `chip` and an additional query parameter called `dale` as empty the URI would be `/disney?name=chip&dale`.
```

File: content/http-handler-abi.md [hunk 5]
```diff
@@ -781,7 +781,10 @@ method to "POST".
 ;;
 ;; Note: The host should return "/" instead of empty for a request with no URI.
 ;;
-;; Note: The URI may include query parameters.
+;; Note: The URI may include query parameters. Despite the path and query being
+;; set as raw, it will write the URI encoded (both path and query parameters).
+;; For example, a set_uri called with `/v1.0/hi?name=kung fu panda` will be writen
```
Comment (anuraaga) [inline content/http-handler-abi.md:786]: See below

**Comment by jcchavezs** [comment]
PTAL @anuraaga @codefromthecrypt

File: content/http-handler-abi.md [hunk 6]
```diff
@@ -808,7 +810,8 @@ below, and the `len` of the URI would be returned:
 ;; set_uri overwrites the URI with one read from memory, e.g.
 ;; "/v1.0/hi?name=panda".
 ;;
-;; Note: The URI may include query parameters.
+;; Note: The URI may include query parameters. The host MUST ensure URI passed
```
Comment (codefromthecrypt) [inline content/http-handler-abi.md:813]: is passed the right word?  below we say "A host who fails to set the URI will trap"...

if we are saying the guest must pass this encoded, we can clarify note below that "A host who receives an unencoded URI or otherwise fails to set the URI will trap"

Otherwise, maybe this line should be something like.. "Note: ... The guest MAY pass the URI unencoded. When a host receives an unencoded URI, it MUST encode it."

**Review by codefromthecrypt — COMMENTED** [review]
As far as I can see we are leaning towards URI being always set/get encoded. and have some good reasons for that.

I made a comment for both directions, I think that simplest is to drop the idea that a guest can go without a URI encoding logic. put that in the rationale as the side effect, and ensure the notes on set_uri say it will panic if the uri isn't encoded.

sg?

File: content/http-handler-abi.md (hunk 6 above)
Comment (jcchavezs) [reply content/http-handler-abi.md:813]: Interesting. How would the host know the URI is unencoded in either case? I think this takes us to the point where we require the guest to encode the URI and we just trust it is encoded because I don't see a way to determine whether `?name=chip&dale` is encoded or not as the spec does not define how to deal with this syntax see https://www.rfc-editor.org/rfc/rfc3986#page-23

**Comment by jcchavezs** [comment]
> and ensure the notes on set_uri say it will panic if the uri isn't encoded.

I partially changed it, I am still not confident host can determine whether uri isn't encoded and panic. Couldn't find a reliable way to determine such a thing without glitches and doing ugly recursions. See:
- https://stackoverflow.com/questions/2295223/how-to-find-out-if-string-has-already-been-url-encoded
- https://stackoverflow.com/questions/17564837/how-to-know-if-a-url-is-decoded-encoded

File: content/http-handler-abi.md (hunk 6 above)
Comment (codefromthecrypt) [reply content/http-handler-abi.md]: so I think we end up needing to have the guest encode it. so maybe something like my former comment. though if we have no way to tell if it is encoded or not, then the host will have no way to strictly enforce it.

 "the guest MUST encode the entire URI (e.g. path in XX encoding and query in YY encoding)" then later "A host who receives an unencoded URI or otherwise fails to set the URI MAY trap"  the latter thing MAY as if it is true we can't tell if something is encoded then we can't say the host must trap on bad encoding.

**Comment by codefromthecrypt** [comment]
thanks @jcchavezs. we indeed need the spec to be implementable. I made a comment that summarizes as a host is allowed to trap (panic) if it can heuristically tell the URI is not encoded properly. however, we aren't forcing the host to trap due to points you mention (which should be in the RATIONALE.. ex "Why doesn't the host trap on bad URI?"

**Review by anuraaga — APPROVED** [review]

File: content/http-handler-abi.md (hunk 6 above)
Comment (anuraaga) [reply content/http-handler-abi.md]: I'm good with the current text or adding some clarification of possible trap, either's good for me. I guess there are probably other cases that using the ABI in an undefined way can cause panics (e.g. writing a non-utf8 string) so I wonder if we want to explicitly write out all of them or to keep it as a general principle. Though URL encoding may be normal enough (as opposed to non-utf8 string) that it's worth documenting anyways

**Comment by jcchavezs** [comment]
Thanks for the feedback. It's been a long run. I will finish it by tomorrow.

**Comment by jcchavezs** [comment]
@codefromthecrypt PTAL

**Review by codefromthecrypt — APPROVED** [review]
well done!
